engine/telemetry/pubsub.go

```diff
@@ -0,0 +1,826 @@
+	}
+	for {
+		if ps.spanDone[key] {
+			return true
+		}
+		if parent, ok := ps.spanParents[key]; ok && parent.IsValid() {
+			key.SpanID = parent
+		} else {
+			return false
+		}
+	}
+}
+
+func (ps *PubSub) trackSpan(span sdktrace.ReadOnlySpan) {
+	key := spanKey{
+		TraceID: span.SpanContext().TraceID(),
+		SpanID:  span.SpanContext().SpanID(),
+	}
+
+	// slog := slog.With("span", span.Name(), "spanID", span.SpanContext().SpanID(), "parentID", span.Parent().SpanID())
+
+	ps.spanParents[key] = span.Parent().SpanID()
+
+	// slog.ExtraDebug("!!! ClientAnnotator tracking span clients")
+
+	for _, attr := range span.Attributes() {
+		if attr.Key == telemetry.ClientIDAttr {
+			// slog.ExtraDebug("!!! > trackSpan tracking client ID", "value", attr.Value.AsString())
+			ps.spanClients[key] = attr.Value.AsString()
+			return
+		}
+	}
+
+	// slog.ExtraDebug("!!! trackSpan did not find client ID")
+}
+
+type ClientMetadata struct {
+	// ClientID is unique to each client. The main client's ID is the empty string,
+	// any module and/or nested exec client's ID is a unique digest.
+	ClientID string `json:"client_id"`
+}
+
+const (
+	// NB: a bit odd to move this here, but feels less risky than duplicating
+	ClientMetadataMetaKey = "x-dagger-client-metadata"
+)
+
+func ClientMetadataFromContext(ctx context.Context) (*ClientMetadata, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("failed to get metadata from context")
+	}
+	clientMetadata := &ClientMetadata{}
+	if err := decodeMeta(md, ClientMetadataMetaKey, clientMetadata); err != nil {
+		return nil, err
+	}
+	return clientMetadata, nil
+}
+
+func decodeMeta(md metadata.MD, key string, dest interface{}) error {
+	vals, ok := md[key]
+	if !ok {
+		return fmt.Errorf("failed to get %s from metadata", key)
+	}
+	if len(vals) != 1 {
+		return fmt.Errorf("expected exactly one %s value, got %d", key, len(vals))
+	}
+	jsonPayload, err := base64.StdEncoding.DecodeString(vals[0])
+	if err != nil {
+		return fmt.Errorf("failed to base64-decode %s: %w", key, err)
+	}
+	if err := json.Unmarshal(jsonPayload, dest); err != nil {
+		return fmt.Errorf("failed to JSON-unmarshal %s: %w", key, err)
+	}
+	return nil
+}
+
+func (ps *PubSub) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ps.mux.ServeHTTP(w, r)
+}
+
+func (ps *PubSub) TracesHandler(rw http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		slog.Warn("error reading body", "err", err)
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var req coltracepb.ExportTraceServiceRequest
+	if err := proto.Unmarshal(body, &req); err != nil {
+		slog.Error("error unmarshalling request", "err", err)
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if err := ps.Spans().ExportSpans(r.Context(), telemetry.SpansFromPB(req.ResourceSpans)); err != nil {
+		slog.Error("error exporting spans", "err", err)
+		rw.WriteHeader(http.StatusInternalServerError)
```

Comment:
nit: not sure if useful but if you do `http.Error(rw, err, http.StatusInternalServerError)` the error message gets sent back to the client to as the response body, which can make debugging easier provided it's printed. May be irrelevant here though, not sure.